File: server/src/models/ProductImage.ts
import {
  AfterSave,
  AllowNull,
  BelongsTo,
  Column,
  DataType,
  Default,
  ForeignKey,
  IsUUID,
  Model,
  PrimaryKey,
  Table,
  Not,
} from "sequelize-typescript";
import { Transaction } from "sequelize/types";
import { Field, ObjectType } from "type-graphql";
import { Product } from ".";

@ObjectType()
@Table({
  timestamps: true,
  underscored: true,
})
export class ProductImage extends Model {
  @Field()
  @IsUUID(4)
  @PrimaryKey
  @Default(DataType.UUIDV4)
  @Column
  id: string;

  @Field()
  @AllowNull(false)
  @Column
  url: string;

  @Field({ nullable: true })
  @Column
  title: string;

  @Field({ nullable: true })
  @Column
  description: string;

  @Field()
  @AllowNull(false)
  @Default(false)
  @Column
  primary: boolean;

  @AfterSave
  static async ensurePrimary(
    instance: ProductImage,
    { transaction }: { transaction: Transaction }
  ) {
    if (instance.primary) {
      await ProductImage.update(
        { primary: false },
        {
          where: { productId: instance.productId, id: Not(instance.id) },
          transaction,
        }
      );
    } else {
      const quantity = await ProductImage.count({
        where: { primary: true },
        transaction,
      });
      if (!quantity) {
        throw new Error("One image must be primary.");
      }
    }
  }

  @IsUUID(4)
  @ForeignKey(() => Product)
  @AllowNull(false)
  @Column
  productId: string;

  @BelongsTo(() => Product)
  product: Product;

  @Field()
  createdAt: Date;

  @Field()
  updatedAt: Date;
}

File: go-server/internal/models/product_image.go
package models

import (
	"errors"

	"gorm.io/gorm"
)

func init() {
	registerModel(&ProductImage{})
}

type ProductImage struct {
	Base

	URL         string `gorm:"not null" json:"url"`
	Title       string `json:"title,omitempty"`
	Description string `json:"description,omitempty"`
	Primary     bool   `gorm:"column:primary;not null;default:false" json:"primary"`

	ProductID string   `gorm:"not null;index" json:"-"`
	Product   *Product `gorm:"foreignKey:ProductID;constraint:OnDelete:CASCADE" json:"-"`
}

// AfterSave mirrors ProductImage.ensurePrimary exactly, apparent bug and
// all: the "at least one primary" check on the else branch counts *every*
// primary image across all products, not just this one's, because the
// Node query never filters by productId either.
func (pi *ProductImage) AfterSave(tx *gorm.DB) error {
	if pi.Primary {
		return tx.Model(&ProductImage{}).
			Where("product_id = ? AND id <> ?", pi.ProductID, pi.ID).
			Update("primary", false).Error
	}

	var count int64
	if err := tx.Model(&ProductImage{}).
		Where(map[string]any{"primary": true}).
		Count(&count).Error; err != nil {
		return err
	}
	if count == 0 {
		return errors.New("One image must be primary.")
	}
	return nil
}
